### ISSUE TITLE
Create Authorization Script to check BCrypt Hash

### ISSUE THREAD

**Issue by afroDC** [GluuFederation/oxAuth#753]
Our default encryption method for stored passwords in OpenLDAP from 3.0.* to 3.1.2 is BCrypt. OpenDJ does not support BCrypt, so we need to develop an authorization script that utilizes a python library that can verify the passwords using this hashing algorithm.

Here is one example:

http://passlib.readthedocs.io/en/stable/index.html

http://passlib.readthedocs.io/en/stable/lib/passlib.hash.bcrypt.html

```
>>> from passlib.hash import bcrypt

>>> # generate new salt, hash password
>>> h = bcrypt.hash("password")
>>> h
'$2a$12$NT0I31Sa7ihGEWpka9ASYrEFkhuTNeBQ2xfZskIiiJeyFXhRgS.Sy'

>>> # the same, but with an explicit number of rounds
>>> bcrypt.using(rounds=13).hash("password")
'$2b$13$HMQTprwhaUwmir.g.ZYoXuRJhtsbra4uj.qJPHrKsX5nGlhpts0jm'

>>> # verify password
>>> bcrypt.verify("password", h)
True
>>> bcrypt.verify("wrong", h)
False
```

**Comment by yurem** [comment]
We also can try to use this method during user entries migration to LDAP server which not supports BCrypt

**Comment by nynymike** [comment]
![image](https://user-images.githubusercontent.com/3717101/36681182-84a1a1d8-1add-11e8-96ea-7a6dddb4bac6.png)

```
title Bcrypt migration
user->browser: enter password
browser->oxAuth: /authorize
oxAuth->oxAuth: bcrypt password
oxAuth->LDAP: retrieve bcrypted password i.e. {bcrypt}xyzabc
oxAuth->oxAuth: compare bcrypted passwords
opt if match
oxAuth->ldap: store password (use clear text)
ldap->ldap: hash password
ldap->oxAuth:
end
oxAuth->browser: response
```

**Comment by mzico** [comment]
@sahiliamsso : Changing the priority of this issue... couple of customers are waiting for this.

**Comment by colonha** [comment]
Maybe upgrade Opendj to version 3.5 which support Bcrypt?

`OpenDJ directory server now implements a Bcrypt password storage scheme that uses the bcrypt message digest algorithm `

https://bugster.forgerock.org/jira/browse/OPENDJ-2435

**Comment by afroDC** [comment]
@colonha As far as I know, OpenDJ 3.5 is not open source,.

**Comment by colonha** [comment]
I read this on their website.

`OpenDJ is free to download, evaluate, and use for developing your applications and solutions. You can obtain and modify the source code to build your own version. ForgeRock offers training and support subscriptions to help you get the most out of your deployment.`

**Comment by afroDC** [comment]
@colonha can you link to the source for OpenDJ 3.5 or greater?

**Comment by colonha** [comment]
I tried but a subscriptions is required. Since they claim you can obtain and modify by your own i'm trying to contact them.

`This version can only be downloaded with an active subscription`

**Comment by nynymike** [comment]
Another option, I added an issue to [gluu-opendj](https://github.com/GluuFederation/gluu-opendj/)

[Add Bcrypt](https://github.com/GluuFederation/gluu-opendj/issues/2)

**Comment by sahiliamsso** [comment]
I have verified [this](https://github.com/afroDC/Dev/blob/master/Gluu/bcryptAuthScript.py) script is working fine. Below are the steps I followed
- Installed Gluu 3.0.2
- Imported 20 test users
- Followed upgrade steps from 3.0.2 to 3.1.3
- Verified all users are available in 3.1.3
- Tried to login with test users -> FAILED authentication
- Enabled Bicrypt Authentication script
- Verified all test Users are able to login
- Disabled Bicrypt authentication scipt
- Verified test users still able to login
- Verified negative test scenarios by providing empty/invalid password

**Comment by yurem** [comment]
@sahiliamsso Can you move it to https://github.com/GluuFederation/oxAuth/tree/master/Server/integrations and close after that?
